[자동차 경주 - 4단계] 리팩토링

Starlight258 opened the pull request:
안녕하세요 진영님 !
자동차 경주 4단계 미션을 완료했습니다 😁

## 궁금한 점
- 도메인 생성시 입력값 검증 책임: `도메인`(Model) vs `Controller` 
   입력값에 대한 검증은 `controller`에서 수행하고, `도메인`은 검증된 값이 들어왔다고 가정하고 생성하도록 작성했습니다!
   도메인에 검증 로직이 들어가면 검증 로직이 복잡할수록 생성자 코드가 길어지기 때문입니다.
   그리고 입력값이 잘못 들어왔을때 컨트롤러 단에서 사용자에게 빠르게 피드백하여 다시 올바른 값을 입력받기 위해 분리했습니다.
   
   그런데 검증 로직을 해당 도메인에 포함하면 자기 자신에 대한 검증 로직이므로 **캡슐화**가 더 잘 될 것 같습니다.
   그리고 입력값이 아니더라도 **잘못된 값에 의해 객체가 생성되는 것을 막을 수 있는 장점**이 있을 것 같습니다.

  진영님은 어떤 방법을 더 선호하시는지 궁금합니다 !

- `Controller` 테스트
   Controller 테스트 작성시에 특정 메서드가 잘 호출되었는지를 확인하는 것을 목표로 두었는데 이 방향이 맞는지 궁금합니다! 
   작성하고보니 굳이 이런 테스트가 필요할지 의문이 생겨서 질문 드립니다 !

- 일급 컬렉션 `Cars`
   리팩토링하면서 자동차 경주에 대한 로직이 캡슐화되어 일급 컬렉션 `Cars`로 구현이 되었는데 일급 컬렉션에 맞게 잘 작성되었는지 궁금합니다!

✅ 구현된 코드에 대한 설명은 `README.md`에서 보실 수 있습니다 ! 

리뷰 부탁드립니다 !! 감사합니다~ 🙇‍♀️

Starlight258 committed:
refactor: 1차 리팩토링

Starlight258 committed:
feat: Generator 인터페이스 도입
- 값을 생성할 수 있는 인터페이스를 도입한다.
- 기존 RandomValueGenerator에 대한 결합도를 줄이고 OCP, DIP 원칙을 지킨다.

Starlight258 committed:
refactor: 일급 컬렉션 도입
- Car 클래스를 포장하는 일급 컬렉션을 도입하여 자동차 경주에 대한 비즈니스 로직을 캡슐화한다.
- cars는 불변 리스트이다.

Starlight258 committed:
rename: 책임 변경에 따른 이름 변경
- Racer 클래스를 InputValidator로 변경
- 기존 Racer의 자동차 경주에 대한 책임이 Cars로 넘어가면서 입력값 검증 로직만 남음

Starlight258 committed:
feat: 생성자 추가

Starlight258 committed:
feat: InputView, OutputView 도입
- GameApplication에서 입력과 출력 로직 분리

Starlight258 committed:
refactor: 변수명에서 타입 제거

Starlight258 committed:
test: 랜덤값을 경계값으로 수정

Starlight258 committed:
refactor: Application에서 InputValidator 바로 호출하도록 수정
- 기존에는 InputView를 통해 InputValidator를 호출하였는데 Application에서 바로 호출하도록 변경했습니다.

Starlight258 committed:
refactor: Application 리팩토링
- Generator 생성자 주입받도록 수정
- 입력값 검증 실패시 예외 던지도록 수정
- 관련있는 로직 그룹화

Starlight258 committed:
refactor: inputValidator 생성자 주입받도록 수정

Starlight258 committed:
refactor: 필요없는 메서드 제거

Starlight258 committed:
refactor: 필요없는 변수 제거

Starlight258 committed:
refactor: 필요없는 메서드 제거 및 파라미터 순서 변경

- validateInput 메서드는 한 줄이므로 굳이 메서드로 추출할 필요가 없어서 제거한다.
- 가독성을 위해 validateInput 메서드의 파라미터 순서를 입력 순서로 변경한다.

Starlight258 committed:
style: 상수 선언 후 줄바꿈 추가

Starlight258 committed:
refactor: Cars 클래스의 메서드 반환값 수정
- 내부 컬렉션이 아니라 Cars 자체를 반환하여 컬렉션에 대한 의존성 줄이기

Starlight258 committed:
refactor: 자동차 경주(Cars) 테스트 분리 및 리팩토링

Starlight258 committed:
refactor: 하드코딩 문자열 포매팅으로 수정

Starlight258 committed:
refactor: 예외 발생시 복구하도록 수정

- 입력값 검증 예외 발생시 올바른 값으로 입력받아 예외를 복구하도록 수정한다.
- 기존 ValidateInput 메서드를 검증 대상에 따라 분리하여 올바르지 않은 값만 다시 검증할 수 있도록 수정한다.

Starlight258 committed:
refactor: 예외 메세지를 OutputView로 이동

- 예외 메세지를 InputValidator에서 OutputView로 옮겨 출력 로직에 대한 관심사를 검증 로직으로부터 분리한다.
- InputValidatorTest : 줄바꿈 로직 변경으로 인한 코드 수정이다.

Starlight258 committed:
docs: README.md 업데이트

- 구현 변경에 따른 README 업데이트

Starlight258 committed:
docs: README 줄바꿈 추가

Starlight258 committed:
refactor: 메서드 순서 변경

Starlight258 committed:
test: 필요없는 Nested 클래스 제거

Starlight258 committed:
refactor: mvc 패턴으로 리팩토링

- 패키지명을 mvc 패턴에 따라 변경한다.
- Application에서 입출력(view)와 domain 메서드 호출 부분을 controller로 이전한다.
- InputValidator를 Validator로 이름을 수정하여 기능을 빠르게 파악할 수 있다.
- Validator를 controller 패키지로 이동하여 입력값 검증은 controller로 이전한다.

Starlight258 committed:
rename: 패키지명 수정

- GameApplication -> Application 이름 변경
- RacingGame -> CarRacingController 이름 변경
- generator 패키지는 model 패키지로 이동

Starlight258 committed:
refactor: cars의 position 추출 로직 cars로 이전

Starlight258 committed:
test: 컨트롤러 테스트 추가

Starlight258 committed:
test: Validator 테스트에 성공 테스트 추가

Starlight258 committed:
fix: 랜덤값 생성 에러 해결
- 랜덤값이 8 이하로 생성되는 문제를 해결한다.

Starlight258 committed:
test: RandomValueGenerator에 대한 테스트 생성

Starlight258 committed:
remove: 줄바꿈 삭제

Starlight258 committed:
docs: README 업데이트

Starlight258 committed:
refactor: 필요없는 코드 제거

- cars에 대한 when 코드는 필요가 없다. 랜덤 값에 의해 이미 우승자를 결정하기 때문이다.

Starlight258 committed:
test: Controller 테스트에서 view 로직 검증 제거

- Controller의 책임은 입력을 받고 적절한 메서드를 호출하여 결과를 뷰에 전달하는 것이다. 우승자가 잘 출력되는지는 view에 대한 책임이므로 제거한다.

Starlight258 committed:
test: 패키지명 추가

- 테스트 코드 패키지를 애플리케이션 코드에 맞춰 추가

jinyoungchoi95 commented on src/main/java/org/duckstudy/controller/CarRacingController.java:
> 입력값에 대한 검증은 controller에서 수행하고, 도메인은 검증된 값이 들어왔다고 가정하고 생성하도록 작성했습니다!
> 도메인에 검증 로직이 들어가면 검증 로직이 복잡할수록 생성자 코드가 길어지기 때문입니다.
> 그리고 입력값이 잘못 들어왔을때 컨트롤러 단에서 사용자에게 빠르게 피드백하여 다시 올바른 값을 입력받기 위해 분리했습니다.
> 
> 그런데 검증 로직을 해당 도메인에 포함하면 자기 자신에 대한 검증 로직이므로 캡슐화가 더 잘 될 것 같습니다.
> 그리고 입력값이 아니더라도 잘못된 값에 의해 객체가 생성되는 것을 막을 수 있는 장점이 있을 것 같습니다.

domain에서 해요. controller에서 fail fast를 해야하는 상황이라면 어쩔 수 없는데, domain에서 검증해야하는 비즈니스로직이라면 controller에 중복이 되더라도 domain에는 무조건 넣습니다.

그래야 안정적으로 생성할 수 있는 domain이 지금 controller가 아니더라도 다른곳에서도 쓰일 수 있어요. 그래서 domain에 필수적으로 넣는거고 fail fast의 상황이 필요하다면 controller에 더 두기도 하는데 개인적으론 fail fast보단 중복코드가 더 불필요하다고 생각해서 domain에만 넣는 편입니다. (fail fast를 해도 io작업이 있지 않는 이상 그렇게 성능차이가 나는것도 아니라서요.)

Starlight258 replied:
그렇군요 !! 검증 로직은 도메인에 두는게 더 낫겠네요! 맞춰서 수정해보겠습니다 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/controller/CarRacingController.java:
```diff
@@ -0,0 +1,57 @@
+import org.duckstudy.view.InputView;
+import org.duckstudy.view.OutputView;
+
+public class CarRacingController {
```
> Controller 테스트 작성시에 특정 메서드가 잘 호출되었는지를 확인하는 것을 목표로 두었는데 이 방향이 맞는지 궁금합니다!
> 작성하고보니 굳이 이런 테스트가 필요할지 의문이 생겨서 질문 드립니다 !

현재 미션에서는 controller에 대한 테스트는 굳이 하지 않으셔도 됩니다. view와 실행 로직을 이어주는 역할로써만 두고 나머지 기능들은 모두 도메인에서 테스트해주시면 되어요.

Starlight258 replied:
그렇군요 !! 알겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/generator/Generator.java:
```diff
@@ -0,0 +1,5 @@
+package org.duckstudy.model.generator;
+
+public interface Generator {
```
추상화 잘 해주셨네요. 무슨 Generator인지도 명시해주시면 좋을 것 같아요.

Starlight258 replied:
이름이 너무 추상적이긴 하네요! 
`RandomValueGenerator`로 인터페이스 이름을 수정하고 구현체는 `DefaultRandomValueGenerator`로 수정하겠습니다 ~!

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/car/CarTest.java:
```diff
@@ -1,47 +1,43 @@
-        randomValueGenerator = mock(RandomValueGenerator.class);
-        carMover = new CarMover(randomValueGenerator);
-        car = new Car("Test Car");
+        this.generator = mock(RandomValueGenerator.class);
```
`Generator`로 추상화를 해주셨으니 `mock`에 대한 의존성을 모두 지우고 추상화를 활용해서 테스트를 작성하는 방법을 고민해주시면 좋을 것 같아요.

Starlight258 replied:
넵..! mock은 최소화로 사용하는게 좋을 것 같아요. 
인터페이스에 구현체 만들어서 테스트 작성해 보겠습니다 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private List<Car> createCars(String[] carNames, Generator generator) {
+        List<Car> cars = new ArrayList<>();
+        for (String carName : carNames) {
+            cars.add(new Car(carName, generator));
+        }
+        return List.copyOf(cars);
```
```suggestion
        return Arrays.stream(carNames)
                .map(name -> new Car(name, generator))
                .toList();
```
로 표현될 수 있어요.

Starlight258 replied:
stream을 이용하면 더 간단하네요 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public Cars(String[] carNames, Generator generator) {
+        this.cars = createCars(carNames, generator);
```
생성자는 보통 안정적인 객체를 보장하는 주생성자 밑에 여러 부생성자를 두는것이 좋아요. 현재는 15~17의 생성자가 된다면,
```suggestion
        this(createCars(carNames, generator));
```
가 될 수 있겠네요.

Starlight258 replied:
그렇군요 ! createCars가 멤버 메서드여서 this 전에 호출이 불가능하니 this 안에 createCars 메서드 내용을 넣어서 수정하겠습니다 !!
```java
public Cars(String[] carNames, RandomValueGenerator randomValueGenerator, OutputView outputView) {
      this(Arrays.stream(carNames)
              .map(carName -> new Car(carName, randomValueGenerator, outputView))
              .collect(Collectors.toList()));
  }
```

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+    public Cars playAndGetWinners(int repetitionNum, OutputView outputView) {
+        for (int i = 0; i < repetitionNum; i++) {
+            moveAll();
+            outputView.printMessage(getPositions());
```
`View`에 대한 의존성은 domain 에서 모두 제거해주세요.

Starlight258 replied:
관심사의 분리를 지켜서 코드 수정하겠습니다 ..! 감사합니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private Cars calculateWinners() {
+        int maxPosition = cars.stream()
+                .mapToInt(Car::getPosition)
+                .max()
+                .orElse(INITIAL_POSITION);
```
1. 메소드로 빼면 어떨까요?
2. max를 계산하지 못하는 상황이라면 0을 내보내는 것이 아닌 예외를 발생시켜야하지 않을까요?

Starlight258 replied:
메서드로 빼고 max를 구하지 못할 경우에는 예외 발생시키도록 하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+                .orElse(INITIAL_POSITION);
+
+        ArrayList<Car> winners = cars.stream()
+                .filter(car -> car.getPosition() == maxPosition)
```
getter를 사용해서 로직을 호출하는 것은 객체끼리의 메시지를 내보내는 형태는 아니에요.

`Car`에 maxPosition인지에 대해서 메시지를 내보내는 형태로 바꾸어보면 어떨까요?

Starlight258 replied:
그 방법이 더 객체지향에 적합하네요 !!  
getter로 객체의 데이터를 직접 접근하지 않고, 객체가 직접 판단하도록 수정하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+                .max()
+                .orElse(INITIAL_POSITION);
+
+        ArrayList<Car> winners = cars.stream()
+                .filter(car -> car.getPosition() == maxPosition)
+                .collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
+        return new Cars(winners);
```
```suggestion
        List<Car> winners = cars.stream()
                .filter(car -> car.getPosition() == maxPosition)
                .toList();
        return new Cars(winners);
```
여도 되지 않을까요? `ArrayList`는 추상화된 `List`와 다른 구현체에요. `추상화에 의존해야지, 구체화에 의존하면 안된다`라는 말이 왜 나왔을까요?

Starlight258 replied:
앗 맞습니다,, 구체 클래스에 의존하면 코드 변경 가능성이 높아집니다.. 수정하겠습니다 ! 😭

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+        return cars.stream()
+                .map(Car::getName)
+                .collect(Collectors.joining(", "));
+    }
```
`,`를 구분자로 출력하는 것은 화면의 요구사항이에요. name을 내보내도록하고 `,`처리는 view에서 해주세요.

Starlight258 replied:
도메인에 view 로직이 포함되어 있었네요,, 수정하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,73 @@
+                .map(car -> car.getName() + " : " + "-".repeat(car.getPosition()))
+                .collect(Collectors.joining("\n"))
+                .concat("\n");
+    }
```
위와 마찬가지입니다.

Starlight258 replied:
넵 데이터만 반환하도록 수정하겠습니다~!

jinyoungchoi95 commented on src/main/java/org/duckstudy/controller/Validator.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Arrays;
+import org.duckstudy.view.OutputView;
+
+public class Validator {
```
`Validator`는 단순히 vaildation처리만 해서 exception을 내보내도록하고 `OutPutView`는 굳이 메시지에 대한 처리를 하지 않아도 되지 않을까요?
- Validator : 값에 대한 예외처리
- OutputView : 화면에 대한 출력처리

Starlight258 replied:
예외 메세지도 OutputView에서 관리하는게 낫지 않을까라고 생각했는데, 검증 예외 메세지는 검증 코드에 포함시키는게 오히려 응집도를 높일 것 같네요 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/controller/Validator.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public int validateRepetitionNum(int repetitionNum) {
+        if (repetitionNum <= 0) {
```
0은 상수로 표현될 수 있겠네요.

Starlight258 replied:
앗 ,, 숫자가 포함되어있었네요 수정하겠습니다 !!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 명지님 😃

몇가지 코멘트남겨두었으니 확인부탁드려요.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: carNames 검증 로직 도메인으로 이전

- 자동차 이름 검증은 자동차 도메인의 책임이므로 Car 클래스로 이전한다.
- 자동차 이름을 검증하는 테스트 코드도 CarTest로 이전한다.

Starlight258 committed:
refactor: repetitonNum 검증 로직 controller로 이전

- 반복 횟수는 Car의 특성이 아니므로 Cars에서 검증을 수행하는 것이 적절하지 않다.
- controller로 반복 횟수 검증 로직을 이전하고 Validator 클래스는 제거한다.

Starlight258 committed:
test: 반복 횟수 검증 테스트 재작성

- 검증 로직이 Controller로 이전하였으므로 그에 맞춰 테스트코드를 이동한다.
- 반복 횟수가 유효하지 않을때 다시 입력받는지 확인하는 테스트 코드를 추가한다.

Starlight258 committed:
refactor: 에러 메세지 수정

Starlight258 committed:
refactor: 반복 횟수 검증 로직 Cars로 이동

- Cars는 자동차 경주 로직이 캡슐화되어있으므로, 반복 횟수 검증 로직도 Cars로 이동한다.
- 테스트 코드도 함께 이동하고 그에 맞춰 수정한다.

Starlight258 committed:
test: 테스트 설명 수정

Starlight258 committed:
rename: 랜덤값 생성 인터페이스, 구현체 이름 변경

- Generator -> RandomValueGenerator
- RandomValueGenerator -> DefaultRandomValueGenerator

Starlight258 committed:
docs: README 업데이트

Starlight258 committed:
remove: Controller 테스트 삭제

Starlight258 committed:
test: CarTest에서 mock 삭제

- mock을 삭제하고 인터페이스에 구현체를 할당한다.

Starlight258 committed:
refactor: 중첩 클래스 내부에 필드 선언하기

- 전역 변수가 아닌 중첩 클래스 내부에 필드를 선언하여 테스트의 독립성을 높인다.

Starlight258 committed:
refactor: domain에서 view 의존성 제거

- 관심사의 분리를 지켜서 Car, Cars에서 OutputView에 대한 의존성을 제거한다.
- controller만 view에 대한 의존성을 갖도록 수정한다.
- 변경된 코드에 맞춰 테스트 코드를 수정한다.

Starlight258 committed:
refactor: domain에 출력 로직(view) 분리

- 도메인에 화면 출력 로직이 포함되지 않게 분리한다.
- 화면 출력 로직은 OutputView로 이동한다.

Starlight258 committed:
refactor: Car에 isWinner 메서드 추가

- 객체가 자신의 상태를 판단하게 한다. (tell, don't ask)

Starlight258 committed:
test: 자동차 결과 테스트 재작성

- mock을 제거하고 실제 경주 로직을 테스트하도록 수정한다.

Starlight258 commented:
@jinyoungchoi95
리뷰해주신 사항들 덕분에 많이 배웠습니다 🙇‍♀️
감사합니다 ! 😁

Starlight258 committed:
refactor: 예외 메세지 하드코딩 제거

Starlight258 commented:
@jinyoungchoi95
리뷰해주신 사항들 덕분에 많이 배웠습니다 🙇‍♀️
감사합니다 ! 😁

Starlight258 committed:
refactor: 메서드명 변경 및 상수 위치 변경

Starlight258 committed:
test: 테스트에 경계값 추가

Starlight258 committed:
style: 줄바꿈 제거

Starlight258 committed:
feat: Application에 리소스 해제 코드 추가

Starlight258 committed:
docs: README 업데이트

Starlight258 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,63 @@
+                .collect(Collectors.toList()));
+    }
+
+    public int validateRepetitionNum(int repetitionNum) {
+        if (repetitionNum <= EXCLUSIVE_MIN_REPETITION_NUM) {
+            throw new IllegalArgumentException(String.format("반복 횟수는 %d보다 커야 합니다.\n", EXCLUSIVE_MIN_REPETITION_NUM));
+        }
+        return repetitionNum;
+    }
```
`repetitionNum`은 Car의 특성이 아니지만 Cars에 게임 로직을 모아두었기 때문에 `repetitionNum 검증 코드`까지 포함시켰습니다. 
Cars에 해당 검증 코드를 포함하는게 좋을지, 아니면 다른 곳으로 이동하는게 좋을지 의견 여쭤봐도 될까요 ??

jinyoungchoi95 replied:
일급컬랙션 `Cars`의 관심사는 `Car`이라는 컬랙션에 대한 관심이기 때문에 단순히 반복횟수에 대한 validation은 어색한 것 같아요. 별도의 value object와 같은 클래스로 분리하는게 좋을 것 같네요.

Starlight258 replied:
맞습니다 !! 값 객체로 분리하는 방법이 있었네요 ! 반영하겠습니다 ! 감사합니다~

Starlight258 committed:
refactor: 필요없는 코드 제거

Starlight258 committed:
feat: AppConfig 추가

- Application에서 객체 생성 및 의존성 주입 책임 분리, 실행 책임만 갖도록 한다.

Starlight258 committed:
refactor: 테스트와 Application의 설정 환경 일치시키기

- AppConfig로부터 의존성을 불러와 테스트 환경과 Application 운영 환경을 일치시킨다.
- 테스트도 추상화된 인터페이스에만 의존할 수 있어 DIP 원칙을 지킬 수 있다.

Starlight258 committed:
rename: AppConfig 파일 config 패키지로 이동

Starlight258 committed:
docs: README 업데이트

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Cars.java:
```diff
@@ -0,0 +1,63 @@
+    private final List<Car> cars;
+
+    public Cars(List<Car> cars) {
+        this.cars = List.copyOf(cars);
```
`copyOf`는 어디까지 복사해서 넣어주나요? 복사를 할 때 어느정도로 복사하는것이 적절할까요?

https://ksh-coding.tistory.com/77

Starlight258 replied:
`List.copyOf`는 새로운 리스트를 만들어 객체들을 복사하고, `Collections.unmodifiableList`는 원본 리스트의 참조를 반환하지만 read-only로 반환합니다. 
`copyOf`를 사용한 이유는 원본 리스트가 변경될 경우 getter로 얻은 복사본 리스트가 함께 변경될 수 있기 때문에, 참조를 끊어주기 위해 사용했습니다.

하지만 원본 리스트가 변경될 경우 복사한 리스트와 상태가 달라지므로 데이터 정합성 측면에서 문제가 발생할 것 같아요. 
그리고 수정을 막는것이 목적이므로 굳이 `List.copyOf`로 불필요하게 리스트 생성을 하는 것보다 `Collections.unmodifiableList`로 원본 리스트 참조는 유지하되 read-only로 반환하는게 좋을 것 같아요.

만약 Cars 외부에서 복사한 리스트의 내부 객체 요소(Car)에 접근해 변경하는 작업이 수행될 경우, `deep copy`가 아니므로 원본 리스트의 객체도 영향이 가서 위험하지 않을까 라는 생각도 했는데요,,,... 
하지만 일급 컬렉션의 특성상 리스트의 내부 객체 요소 변경 작업은 컬렉션 안에서만 이루어져 외부에서 변경되지 않음을 전제조건으로 하기 때문에 괜찮을 것 같습니다 !

jinyoungchoi95 replied:
외부에서 리스트참조값에 대한 정보를 알고있어이를 끊을수있는건 맞지만 방어적복사의 관점에서 Car 각각에대한 접근이 가능하기때문에 위험할수도 있어요. 이들을 방어적복사로 일일히 모두 복사하기 힘들기때문에 최대한 객체를 불변하게 만들어 객체접근으로 상태를 변경하지못하게 하려는 이유도 있고요 (물론 언어에서 기능을 제공하는 언어도 있습니다)

방어적복사를 구현하실필요는없으나 깊은복사라고해도 결국 참조값접근은 가능하다는점만 알아두시면 좋을거같아요.

Starlight258 replied:
그렇군요! 내부 객체의 불변성이 보장되지 않는다면 깊은 복사를 해서 내부 요소들의 참조가 끊기더라도 여전히 객체가 변경이 가능한 문제가 있네요 !!
Car은 움직일때마다 위치 특성이 달라지므로 불변 객체로 생성한다면 이동시마다 새로운 객체를 반환해야하는데 오히려 그 방법이 안전성이 높고 협업시 실수할 가능성이 낮아질 것 같아요.
감사합니다 ~!!

jinyoungchoi95 commented on src/main/java/org/duckstudy/controller/CarRacingController.java:
```diff
@@ -0,0 +1,60 @@
+
+        for (int i = 0; i < repetitionNum; i++) {
+            cars.moveAll();
+            outputView.printIntermediateResult(cars.getAllNames(), cars.getAllPositions());
```
이렇게 들어오면 List의 index를 각각 따로 사용할 수도 있으니 `Cars`는 `List<Car>`만 내보내도록하고 `List<Car>`를 돌면서 필요한 값을 받아 출력하는것이 낫지 않을까요?

Starlight258 replied:
아 그렇겠네요! 리스트만 인자로 주면 출력시 인덱스 실수가 날 확률이 적고 더 명확할 것 같습니다 !!

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/car/CarsTest.java:
```diff
@@ -0,0 +1,65 @@
+    @BeforeEach
+    void beforeEach() {
+        AppConfig appConfig = new AppConfig();
+        RandomValueGenerator randomValueGenerator = appConfig.randomValueGenerator();
```
`Random`구현체를 가져오는 것보다는 `CarTest`처럼 커스텀된 구현체를 넣는게 더 적절하지 않을까요?

Starlight258 replied:
테스트를 `AppConfig`로부터 의존성을 불러와서 실행시키면 현재 운영 환경에서 로직이 잘 돌아가는지 테스트할 수 있어서 그렇게 작성했습니다 ! 
그런데 생각해보면 테스트 환경이 `AppConfig` 설정에 의존하게 되어 테스트의 독립성이 깨지겠네요 ! 🧐

테스트에서 환경 설정은 분리하고 테스트 대상 클래스의 동작을 검증하는데에만 집중하도록 하겠습니다 !
감사합니다 !!

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/car/CarsTest.java:
```diff
@@ -0,0 +1,65 @@
+        void calculateWinnerWhenRaceIsOver() {
+            int repetitionNum = 5;
+
+            for (int i = 0; i < repetitionNum; i++) {
+                cars.moveAll();
+            }
+            Cars winners = cars.calculateWinners();
+
+            assertThat(winners.getAllNames()).isSubsetOf(CAR_NAMES);
```
이러면 winner를 계산할 때 예를들어 position이 1, 0, 1인 경우를 테스트하진 못하는데요. `Cars`를 생성할 때 원하는 위치를 가지는 `List<Car>`들을 주입한다면 `moveAll`도 하지 않고 `calculateWinners`를 테스트할 수 있지 않을까요?

Starlight258 replied:
검증 범위를 너무 크게 잡아서 제대로 검증이 안되는 문제가 있었네요! 수정하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/generator/RandomValueGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package org.duckstudy.model.generator;
+
+public interface RandomValueGenerator {
```
숫자를 반환하기 때문에 `NumberGenerator`까지가 적절하지 않을까 싶어요. 현재 이 인터페이스는 "숫자를 만드는 행위"를 추상화한 것이니까요.

Starlight258 replied:
인터페이스 이름을 어떻게 작성해야할지 고민되었는데 반환값 기준으로 작성하는 좋은 방법이 있었네요 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/generator/DefaultRandomValueGenerator.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.Random;
+
+public class DefaultRandomValueGenerator implements RandomValueGenerator {
```
위에 이어나가면 오히려 여기선 `RandomValueGenerator`가 더 적절하겠네요.

Starlight258 replied:
피드백 감사합니다 ! 구현체에서는 랜덤값을 생성한다는 의미가 드러나는게 필요하겠네요 !!
그런데 현재 구현에서는 숫자값만 반환하므로 제안해주신 인터페이스 이름인 `NumberGenerator`에 맞춰
`RandomNumberGenerator`가 더 명확한 것 같은데 어떻게 생각하시나요?

jinyoungchoi95 replied:
네 좋은거같아요.

jinyoungchoi95 commented on src/main/java/org/duckstudy/config/AppConfig.java:
```diff
@@ -0,0 +1,23 @@
+import org.duckstudy.view.InputView;
+import org.duckstudy.view.OutputView;
+
+public class AppConfig {
```
`Config`는 현재 미션상에선 과한 스펙이기 때문에 그냥 Application에서 생성하여쓰셔도됩니다.

Starlight258 replied:
앗 그렇군요 ! 현재 미션에서는 굳이 Config 클래스까지 만들 필요가 없었던 거네요 !
지적 감사합니다~ 상황에 맞게 적절하게 설계해보겠습니다 !

jinyoungchoi95 reviewed (changes requested):
안녕하세요 명지님 😃

몇가지 코멘트남겨두었으니 확인부탁드려요.

감사합니다🙇‍♂️

Starlight258 committed:
feat: 반복 횟수 VO 생성

- 반복 횟수 VO인 RepetitionCount를 생성하여 Cars로부터 반복 횟수 검증 코드를 분리한다.
- VO를 record로 생성하여 불변(final)을 지키고 boilerplate 코드를 제거한다.
- 테스트 코드를 분리한다.
- RepetitionNum -> RepetitionCount로 변수명을 변경하여 명확하게 한다.

Starlight258 committed:
refactor: 일급 컬렉션 리스트 복사 copyOf -> unmodifiableList로 변경

- 리스트 수정을 막는 것이 목적이므로, 새로운 리스트를 생성하는 copyOf보다 원본 리스트를 read-only로 반환하는 unmodifiableList를 사용한다. (불필요한 메모리 오버헤드)
- 외부에서 리스트에 직접 접근해 수정하는 것을 막고, 수정 시에는 일급 컬렉션의 메서드를 사용하도록 한다.
- 이를 통해 캡슐화와 응집도를 높인다.

Starlight258 committed:
refactor: 필요없는 코드 삭제

- record에서 equals 메서드를 만들어주므로 해당 코드는 필요가 없다.

Starlight258 committed:
refactor: 인덱스 실수 방지

- 이름과 위치 리스트를 각각 인자로 제공하는 것보다 List<Car>를 제공하여 인덱스 실수를 방지하고 코드를 더 명확하게 한다.

Starlight258 committed:
fix: 테스트에서 AppConfig 의존성 제거, 우승자 검증 코드 수정

- 테스트가 AppConfig의 환경 설정에 따라 결과가 달라지므로 AppConfig 의존성을 제거한다.
- 우승자가 예상대로 계산되었는지 검증할 수 있도록 수정한다.

Starlight258 committed:
style: 코드 포맷 변경

Starlight258 committed:
rename: generator 패키지 파일 이름 변경

- RandomValueGenerator -> NumberGenerator
- DefaultRandomValueGenerator -> RandomNumberGenerator

Starlight258 committed:
remove: AppConfig 삭제

- 의존 관계가 바뀌지 않으므로 AppConfig는 오버스펙이다.

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/car/Car.java:
```diff
@@ -0,0 +1,47 @@
+
+    public void move() {
+        int randomValue = numberGenerator.generateNumber();
+        if (randomValue >= MIN_MOVEMENT_VALUE) {
```
움직임 자체에 대한 책임을 넘겨서 `MovableStratey`로 두고 4이상이면 true을 반환하는 구현체를 두는 설계도 생각해볼수있습니다.

Starlight258 replied:
전략 패턴을 사용하면 차가 움직이는 방식을 변경할 수 있겠군요 !! 후에 이동방식을 교체할 수 있는 좋은 방법이네요!
한번 적용해보겠습니다 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/generator/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public int generateNumber() {
+        return new Random().nextInt(EXCLUSIVE_MAX_VALUE);
```
계속 `Random`을 생성하는거보다 상수로 두면 불필요한 객체생성을 막을수 있겠네요.

Starlight258 replied:
앗 그러네요! 요청마다 Random 객체가 생성되는 것보다 한번 생성하고 공유하는 것이 좋겠네요! 감사합니다 !

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/car/CarsTest.java:
```diff
@@ -0,0 +1,28 @@
+                new Car("Car2", ()-> 2),
+                new Car("Car3", ()-> 4)
+        ));
+        cars.moveAll();
```
position을 1,0,1로 각각 생성한 객체를 두면 moveAll도 안할수있겠네요.

Starlight258 replied:
이전에는 position을 Car 객체 안에서만 관리해서  position을 생성자에서 지정할 수 없어서 moveAll을 호출했습니다 !
하지만 Car을 불변 객체로 변경한 후 position을 생성자를 통해 지정할 수 있도록 바뀌어서, moveAll()을 호출하지 않고도 테스트 가능할 것 같아요 ! 반영하겠습니다 ! 감사합니다~

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/car/CarsTest.java:
```diff
@@ -0,0 +1,28 @@
+
+        Cars winners = cars.calculateWinners();
+
+        assertThat(winners.toList().size()).isEqualTo(2);
```
hasSize를 사용하고, 전체 assertion은 asserAll로 묶는게 좋을거같아요.

Starlight258 replied:
검증 문장이 여러개 있을 경우 assertAll을 꼭 이용하겠습니다 ! 감사합니다 !

jinyoungchoi95 reviewed (approved):
안녕하세요 명지님 😃

현재 미션에서 해야할 학습과 구현을 충분히 하신 것 같아 머지하겠습니다.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: Car 불변 객체로 변경

- Car을 불변 객체로 변경하여 일급 컬렉션 Cars의 리스트 내부 객체 요소 Car에 대해서 불변성을 보장한다.

Starlight258 committed:
feat: Car의 이동방식을 전략 패턴으로 구현

- 자동차가 이동하는 방식이 변경될 수 있음을 고려하여 자동차의 이동방식을 MovableStrategy 인터페이스로 구현하여 분리한다.
- 현재는 랜덤값에 의해 이동하므로 RandomMovableStrategy로 구현하여 의존성을 주입한다.
- 그에 맞춰 테스트 코드도 수정한다.

Starlight258 committed:
refactor: Random 객체를 상수로 두어 요청마다 불필요한 객체 생성 막기

- 랜덤값을 생성할때마다 Random 객체가 생성되므로 상수로 두어 한번만 생성되도록 한다.

Starlight258 committed:
fix: assertAll 사용하여 모든 검증 결과 확인

- assertAll을 이용해 모든 검증에 대한 결과를 확인할 수 있도록 수정한다.

Starlight258 commented:
@jinyoungchoi95
피드백 감사합니다 ! 😁
많이 배우네요 ㅠㅠ 감사해요 !!

Starlight258 committed:
docs: README 업데이트
